feat(server): Add support for PFADD and PFCOUNT

chakaz opened the pull request:
This version does not create sparse-encoded HLLs, however it is fully compatible with such ones created by Redis as it converts them to the dense encoding.

#925

<!--
**Commits Must Be Signed and Your PR title must conform to the conventional commit spec**
  * See: https://github.com/dragonflydb/dragonfly/blob/main/CONTRIBUTING.md
  * Please follow the section on `pre-commit hooks`, a linter will validate before you push

  Example PR Title: <type>(<scope>)!: <description>

  * `type` = bug, chore, feat, fix, docs, build, style, refactor, perf, test
  * `!` = OPTIONAL: signals a breaking change
  * `scope` = Optional when `type` is "chore" or "docs"
  * `description` = short description of the change

Examples:

  * chore(examples): Clarify `docker` usage #120
  * docs(readme): Fix Example Links #121
  * feat(ingest)!: Add new ingest #122
  * fix(ingest): Refactor for loop to list comprehension #123
-->

romange commented on src/server/hll_family.cc:
```diff
@@ -0,0 +1,214 @@
+  if (it.ok()) {
+    string hll;
+    it.value()->second.GetString(&hll);
+    ConvertToDenseIfNeeded(hll);
```
Should we do conversations during reads? 
I would expect that we wont have allocations (string) during counts. Is it possible?

chakaz replied:
The actual common allocation here is the `GetString()` call. I couldn't find a way to get a string's value without allocation (I figured it's due to non-contiguous memory?)
The conversion to dense will only be done in the (less-common?) case in which a string was imported from Redis (or somehow manually crafted) and then used. It's a no-op for operations done within Dragonfly (or for dense strings that originated in Redis as well).

romange replied:
@chakaz please add a TODO: saying that we should support this use-case in compact_object, and to provide 
`string_view GetView(string* scratch) const`  interface there

romange reviewed (commented):
Wow, super cool!

romange commented on src/server/hll_family.cc:
```diff
@@ -0,0 +1,214 @@
+  using CI = CommandId;
+
+  *registry << CI{"PFADD", CO::WRITE, -3, 1, 1, 1}.SetHandler(PFAdd)
+            << CI{"PFCOUNT", CO::WRITE, -2, 1, -1, 1}.SetHandler(PFCount);
```
Should not be write, should it?

chakaz replied:
It actually is a write:

> Note: as a side effect of calling this function, it is possible that the HyperLogLog is modified, since the last 8 bytes encode the latest computed cardinality for caching purposes. So PFCOUNT is technically a write command.
(source: https://redis.io/commands/pfcount/)

chakaz commented on src/server/hll_family.cc:
```diff
@@ -0,0 +1,214 @@
+  }
+}
+
+HllBufferPtr StringToHllPtr(string& hll) {
```
Actually Google reverted this to allow (and even encourage) mutable references:
> Non-optional input parameters should usually be values or const references, while non-optional output and input/output parameters should usually be references (which cannot be null)
(source: https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs)

I am, of course, happy to change it to a pointer if you prefer..

romange replied:
My life was a lie! 

![ezgif com-optimize](https://user-images.githubusercontent.com/3674760/235089642-b50c5141-cf73-4b16-855a-74a8f9095614.gif)

Seriously though, give then the whole project is built with this rule, we won't reverse it now 🤷🏼

chakaz replied:
Sure, will do.

chakaz commented on src/server/hll_family.cc:
```diff
@@ -0,0 +1,214 @@
+}
+
+OpResult<int64_t> PFCountMulti(CmdArgList args, ConnectionContext* cntx) {
+  vector<OpResult<string>> hlls;
+  hlls.reserve(args.size());
+
+  Mutex mu;  // Guards `hlls`
+
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    ArgSlice shard_args = t->GetShardArgs(shard->shard_id());
+    lock_guard guard(mu);
+    ReadValues(t->GetOpArgs(shard), shard_args, hlls);
+    return OpStatus::OK;
+  };
```
That's a good idea!
It would require additional allocations though, so do you think it might be worth it?

romange replied:
Yes, we do the same thing in many operations.

chakaz replied:
Done!

chakaz commented:
> It should be enough to check just after inserts, as it doesn't mutate between reads
It's true that all HLLs created by Dragonfly should always be in the dense format, but in order to be compatible with Redis (like if Redis is the master) we need to support reading strings not crafted by Dragonfly which may use sparse encoding.
Or did I miss your intention? Sorry if so!

romange commented:
You are right that it can happen during the rdb load or using the replication from Redis master.
1. Please add a comment about the case
2. Please add a test to rdb_test as well. This is how I did it. I run redis 6.2, added a couple of items , in this case pf counts, and then shutdown the server which produces dump.rdb. Then I rename it, move it to testdata and use it in rdb_test to show compatibility.  We want both dense and sparse items be present in the rdb, I guess.

chakaz commented:
I'm happy to add the rdb_test, but just making sure before: have you looked at the `WorkWithSparseEncoding` test that I added? I took that string from Redis, and so I'm not sure what additional coverage will rdb_test add?
I guess I could also add a Redis-created dense-encoding string, would that do?

chakaz commented:
Great, done!

dranikpg commented on src/server/hll_family.cc:
```diff
@@ -0,0 +1,220 @@
+
+  OpResult<PrimeIterator> it = db_slice.Find(op_args.db_cntx, key, OBJ_STRING);
+  if (it.ok()) {
+    string hll;
+    // TODO: We should provide a GetView() method in compact object to avoid some allocations:
+    // string_view GetView(string& scratch);
+    it.value()->second.GetString(&hll);
```
We already have `GetSlice()`. We can leave a TODO: rename GetSlice to GetView 🙂

romange replied:
this is the line `if ((taglen_ == 0 && encoded.size() < (1 << 15))) {`

chakaz replied:
Sorry that I missed this line!
Do you mean that I should modify this line globally, for strings which are not HLL? I'm surprised by this, as the current number (2^15) must have been tested rigorously. Is there a way to specify a different value for custom `SetString()` calls? Or did you really mean I should change this value for everyone?

romange replied:
It has not been tested rigorously. It's a stupid over-optimization that I implemented more than a year ago. It really matters only for small strings, for strings of length 1K and more it's not important. So setting 8K limit is absolutely fine.

chakaz commented:
I switched to use `rdb_test` indeed, as explicitly writing the dense-encoded string was just too long (took > 1000 lines!).
In addition, I switched to use `GetSlice()` which should save a bunch of allocations, so nice optimization!

PTAL.

romange commented on src/server/hll_family.cc:
```diff
@@ -0,0 +1,225 @@
+  OpResult<PrimeIterator> it = db_slice.Find(op_args.db_cntx, key, OBJ_STRING);
+  if (it.ok()) {
+    string hll;
+    string_view hll_view = it.value()->second.GetSlice(&hll);
```
Have you checked it does not allocate? Based on the code it should because of small string optimization. You should decrease the constant there to 8k imho

chakaz replied:
I just checked (after modifying the const to be 8k) with GDB, and indeed there is no allocation now. I also checked before modifying the constant, and indeed we fell into the small string optimization, which got us to allocate upon reading.
Nice catch I must say!

romange reviewed (commented):
Looks great!
